Establish structure and process for in-repo release notes (#18427)
bevyengine/bevy: working-migration-guides/README.md

> # Bevy's Release Notes Process

Reviewer: ```suggestion
# Bevy's Migration Guide Process
```

Hi! Did someone add `M-Needs-Migration-Guide` to your PR? If so, you're in the right place.

> Let's talk about how this process works.

Reviewer 2: Future Work: Should consider a CI task to do the inverse too; if the diff includes a new or edited migration guide, the PR should be automatically labelled (or flagged for labelling in a comment).

When we make breaking changes to Bevy, we need to communicate them to users so their libraries and applications can be moved to the new Bevy version.

> To do this, we write and ship a migration guide for every major Bevy version.

Reviewer: Similarly, we could link to <https://bevyengine.org/learn/migration-guides/introduction/> here.

> To do this, we write and ship a migration guide for every major Bevy version.

Author: ```suggestion
To do this, we write and ship a [migration guide](https://bevyengine.org/learn/migration-guides/introduction/) for every major Bevy version.
```

> To avoid a crunch at the end of the cycle as we *write* all of these,
> Bevy asks authors (and reviewers) to write draft release notes as part of the pull requests that make breaking changes.

Reviewer: ```suggestion
To avoid a crunch at the end of the cycle as we *write* all of these,
Bevy asks authors (and reviewers) to write a draft migration guide as part of the pull requests that make breaking changes.
```

> ## What to put in your draft release notes

Reviewer: ```suggestion
## What to put in your draft migration guide
```

A `template.md` file is provided in `bevyengine/bevy/working-migration-guides`: copy-paste that to get started!

Migration guides are intended to briefly communicate:

> - what has been changed?

Reviewer 2: ```suggestion
- what has been changed since the last release?
```
It _should_ go without saying but I forget to do this all the time! It's easy to get tunnel vision and just write a migration for what your PR does to `main`.

- why did we make this breaking change?
- how can users migrate their existing code?

Draft migration guides *do not need to be polished*: it's okay if you're not a native English speaker or aren't a wordsmith.
Editing is easy; we just want to have an expert's view on the questions above.

When writing migration guides, prefer terse, technical language, and be sure to include terms that users might search for.
Migration guides are not read end-to-end: instead, they are navigated via Ctrl+F as the reader follows the compiler errors and bugs.

## Where to put your migration guides

Each major Bevy version (e.g. 0.12, or 2.0) will get its own migration guide.
The draft migration guides are organized into a folder of the same name inside of `bevyengine/bevy/working-migration-guides`.

> When we release our first release candidate for a cycle, these notes are moved from `bevyengine/bevy`, and into `bevyengine/bevy-website`,

Reviewer 2: ```suggestion
When we publish our first release candidate for a cycle, these notes are moved from `bevyengine/bevy`, and into `bevyengine/bevy-website`,
```
Nit: Avoid repetition

> However, it's not always possible to use this attribute, and Bevy does not consider it to be a substitute to a migration guide entry.

Reviewer: This seems like a useful but somewhat unrelated aside. Maybe it should be placed at the end? That way, the sections focusing on the actual migration guide process are first, and the text maybe flows a bit better. With the current order, I felt like this was kinda sandwiched in, and I had to read pretty far to find where the guide should actually go.

## Where to put your migration guides

Each major Bevy version (e.g. 0.12, or 2.0) will get its own migration guide.

> The draft release notes are organized into a folder of the same name inside of `bevyengine/bevy/working-migration-guides`.

Reviewer: ```suggestion
The draft migration guides are organized into a folder of the same name inside of `bevyengine/bevy/working-migration-guides`.
```

If there are *multiple* breaking changes layered on top of each other,
you should edit the existing migration guide, rather than write a new one.

While some brave users live on Bevy's `main` branch, we can trust them to use the draft migration guides and read the PRs in question if needed.

As a result, each draft migration should be given a clear name matching the section title.

> These titles should reflect the name of the old feature that was broken or changed.

Reviewer 2: This implies there may be some PRs which should have multiple migration guides, since a PR might have ramifications that affect multiple seemingly distinct features. Not a critique, just want to highlight that possibility.
